NPB/FT/ft.cpp

```diff
@@ -0,0 +1,1160 @@
+				// 	xout.local(k,j,i+ii).real = y0[j][i].real;
+				// 	xout.local(k,j,i+ii).imag = y0[j][i].imag;
+				// }
+				int offset = k*NX*NY+j*NX;
+				std::copy(&y0[j][0], &y0[j][fftblock], xout.lbegin()+offset+ii);
+				//
+				// The combination of reading in sequential and writing with std::copy
+				// seems to be the fastest with my setup. This definitely needs
+				// further investigation.
+				//
+			}
+		//	if (TIMERS_ENABLED == TRUE) timer_stop(T_FFTCOPY);
+		}
+	}
+	if (TIMERS_ENABLED == TRUE) timer_stop(T_MAX);
+}
+
+/*--------------------------------------------------------------------
+c-------------------------------------------------------------------*/
+
+static void cffts3(int is, int d[3], dash::NArray<dcomplex, 3> &x, dash::NArray<dcomplex, 3> &xout) {
+	/*--------------------------------------------------------------------
+	c-------------------------------------------------------------------*/
+
+	int logd[3];
+
+	for (int i = 0;i < 3; i++) {
+		logd[i] = ilog2(d[i]);
+	}
+
+	if (TIMERS_ENABLED == TRUE) timer_start(T_MAX);
+
+	int yplanes_per_unit = ceil((double) d[1] / dash::size()); //this is effectively a blocked pattern. It could be better to use cyclic for higher unit counts.
+	int myoffset = dash::myid()*yplanes_per_unit;
+
+	for(int j = 0; j < yplanes_per_unit && myoffset+j < d[1]; j++) {
+
+		dcomplex y0[NX][FFTBLOCKPAD];
+		dcomplex y1[NX][FFTBLOCKPAD];
+
+		// dash::Future<dcomplex*> futs[d[2]];
+		// typedef typename dash::Future<dash::GlobIter<dcomplex, dash::BlockPattern<3, (dash::MemArrange)1, long int>, dash::GlobStaticMem<dash::HostSpace>, dash::GlobPtr<dcomplex, dash::GlobStaticMem<dash::HostSpace> >, dash::GlobRef<dcomplex> > > async_write_type;
+		// async_write_type futs_w[d[2]];
+
+		//
+		// Asynchronous copy does not seem to yield anything, at best it is equally good.
+		//
+
+		for (int ii = 0; ii <= d[0] - fftblock; ii+=fftblock) {
+			//	if (TIMERS_ENABLED == TRUE) timer_start(T_FFTCOPY);
+			for (int k = 0; k < d[2]; k++) {
+				// for (int i = 0; i < fftblock; i++) {
+					 //y0[k][i] = x(k,myoffset+j,i+ii);
+				// }
+				int offset = k*NX*NY+(myoffset+j)*NX;
+				dash::copy(x.begin()+offset+ii, x.begin()+offset+ii+fftblock, &y0[k][0]);
+				// futs[k] = dash::copy_async(x.begin()+offset+ii, x.begin()+offset+ii+fftblock, &y0[k][0]);
+			}
+
+			// for (int k = 0; k < d[2]; k++) futs[k].wait();
+
+			//	if (TIMERS_ENABLED == TRUE) timer_stop(T_FFTCOPY);
+			//	if (TIMERS_ENABLED == TRUE) timer_start(T_FFTLOW);
+			cfftz (is, logd[2], d[2], y0, y1);
+			//	if (TIMERS_ENABLED == TRUE) timer_stop(T_FFTLOW);
+			//	if (TIMERS_ENABLED == TRUE) timer_start(T_FFTCOPY);
+			for (int k = 0; k < d[2]; k++) {
+				// for (int i = 0; i < fftblock; i++) {
+					// xout(k,myoffset+j,i+ii) = y0[k][i];
+				// }
+				int offset = k*NX*NY+(myoffset+j)*NX;
+				dash::copy(&y0[k][0], &y0[k][fftblock], xout.begin()+offset+ii);
+				// futs_w[k] = dash::copy_async(&y0[k][0], &y0[k][fftblock], xout.begin()+offset+ii);
```

Comment:
Why not use `copy_async`? I'd guess that allows multiple concurrent data streams, which should be faster...

Reply:
That was my guess as well, but on my computer at home, it didn't help, that's why i commented it out. This can be changed by whoever uses these benchmarks.